bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/Resource.java

```diff
@@ -0,0 +1,609 @@
+    private @Nullable ColorXy color;
+    private @Nullable Alerts alert;
+    private @Nullable Effects effects;
+    private @Nullable @SerializedName("timed_effects") Effects timedEffects;
+    private @Nullable ResourceReference group;
+    private @Nullable List<ActionEntry> actions;
+    private @Nullable Recall recall;
+    private @Nullable Boolean enabled;
+    private @Nullable LightLevel light;
+    private @Nullable Button button;
+    private @Nullable Temperature temperature;
+    private @Nullable Motion motion;
+    private @Nullable @SerializedName("power_state") Power powerState;
+    private @Nullable @SerializedName("relative_rotary") RelativeRotary relativeRotary;
+    private @Nullable List<ResourceReference> children;
+    private @Nullable JsonElement status;
+
+    /**
+     * Constructor
+     *
+     * @param resourceType
+     */
+    public Resource(@Nullable ResourceType resourceType) {
+        if (Objects.nonNull(resourceType)) {
+            setType(resourceType);
+        }
+    }
+
+    /**
+     * Put this resource's control id in the given map of control ids.
+     *
+     * @param controlIds the map of control ids to be updated.
+     * @return this resource instance.
+     */
+    public Resource addControlIdToMap(Map<String, Integer> controlIds) {
+        if (!hasSparseData) {
+            MetaData metadata = this.metadata;
+            controlIds.put(getId(), Objects.nonNull(metadata) ? metadata.getControlId() : 0);
+        }
+        return this;
+    }
+
+    /**
+     * Method that uses reflection to copy fields from an other Resource instance into this instance. If the field in
+     * this instance is null and the same field in the other instance is not null, then the value from the other
+     * instance is copied to this instance.
+     *
+     * Usage: For color light resources, a full DTO requires both a Dimmer and a ColorXy field, but the SSE event DTOs
+     * only provide one or the other field, and never both. This method allows the binding to import the missing field
+     * from a prior cache copy of the resource into itself, in order to create the required full DTO.
+     *
+     * @param other the other resource instance.
+     * @return this instance.
+     */
+    public Resource copyMissingFieldsFrom(Resource other) {
+        for (String fieldName : Set.of("dimming", "color")) {
+            try {
+                Field field = getClass().getDeclaredField(fieldName);
+                Object thisField = field.get(this);
+                Object otherField = field.get(other);
+                if (Objects.isNull(thisField) && Objects.nonNull(otherField)) {
+                    field.setAccessible(true);
+                    field.set(this, otherField);
+                }
+            } catch (IllegalArgumentException | IllegalAccessException | NoSuchFieldException | SecurityException e) {
+            }
+        }
+        return this;
+    }
```

Comment:
Why use reflection here?

```diff
diff --git a/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/Resource.java b/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/Resource.java
index ca196aed6c..5600753133 100644
--- a/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/Resource.java
+++ b/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/Resource.java
@@ -12,11 +12,9 @@
  */
 package org.openhab.binding.hue.internal.dto.clip2;
 
-import java.lang.reflect.Field;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
-import java.util.Set;
 
 import org.eclipse.jdt.annotation.NonNullByDefault;
 import org.eclipse.jdt.annotation.Nullable;
@@ -149,17 +147,11 @@ public class Resource {
      * @return this instance.
      */
     public Resource copyMissingFieldsFrom(Resource other) {
-        for (String fieldName : Set.of("dimming", "color")) {
-            try {
-                Field field = getClass().getDeclaredField(fieldName);
-                Object thisField = field.get(this);
-                Object otherField = field.get(other);
-                if (Objects.isNull(thisField) && Objects.nonNull(otherField)) {
-                    field.setAccessible(true);
-                    field.set(this, otherField);
-                }
-            } catch (IllegalArgumentException | IllegalAccessException | NoSuchFieldException | SecurityException e) {
-            }
+        if (dimming == null) {
+            dimming = other.dimming;
+        }
+        if (color == null) {
+            color = other.color;
         }
         return this;
     }
```
...compiles equally well.

Reply:
Aha. The field DTO's are all `private` to prevent other parts of the binding from accessing them except via the getter/setters. But I forgot that one instance of a class can access the private fields of another instance of the same class. (Perhaps a weakness of the Java private schema.. ??)